// typescript-openapi-router/dist/router/request.js
"use strict";
/**
 * Copyright 2020 Valentin Draganescu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var logger_1 = require("../logger");
var Request = /** @class */ (function () {
    function Request(opts) {
        var _a;
        this.headers = opts.headers;
        this.path = opts.path;
        this.method = opts.method;
        this.body = opts.body;
        this.queryParams = (_a = opts.queryParams, (_a !== null && _a !== void 0 ? _a : this.getQueryParams()));
        this.rawBody = opts.rawBody;
        this.extra = {};
    }
    Request.prototype.getQueryParams = function () {
        logger_1.Logger.log("Getting query params from", this.path);
        if (this.path) {
            var paramPath = this.path.split("?")[1];
            if (paramPath) {
                logger_1.Logger.log("Got param path::", paramPath);
                var params = {};
                var paramPairs = paramPath.split("&");
                logger_1.Logger.log("There are " + paramPairs.length + " params");
                for (var _i = 0, paramPairs_1 = paramPairs; _i < paramPairs_1.length; _i++) {
                    var paramPair = paramPairs_1[_i];
                    var paramPairComponents = paramPair.split("=");
                    params[paramPairComponents[0]] = paramPairComponents[1];
                }
                logger_1.Logger.log("Got query params::", params);
                return params;
            }
        }
        return null;
    };
    return Request;
}());
exports.Request = Request;
