// typescript-openapi-router/dist/router/status-code.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["okay"] = 200] = "okay";
    StatusCode[StatusCode["created"] = 201] = "created";
    StatusCode[StatusCode["accepted"] = 202] = "accepted";
    StatusCode[StatusCode["noContent"] = 204] = "noContent";
    StatusCode[StatusCode["movedPermanently"] = 301] = "movedPermanently";
    StatusCode[StatusCode["found"] = 302] = "found";
    StatusCode[StatusCode["seeOther"] = 303] = "seeOther";
    StatusCode[StatusCode["notModified"] = 304] = "notModified";
    StatusCode[StatusCode["temporaryRedirect"] = 307] = "temporaryRedirect";
    StatusCode[StatusCode["badRequest"] = 400] = "badRequest";
    StatusCode[StatusCode["unauthorized"] = 401] = "unauthorized";
    StatusCode[StatusCode["forbidden"] = 403] = "forbidden";
    StatusCode[StatusCode["notFound"] = 404] = "notFound";
    StatusCode[StatusCode["conflict"] = 409] = "conflict";
    StatusCode[StatusCode["internalServerError"] = 500] = "internalServerError";
    StatusCode[StatusCode["notImplemented"] = 501] = "notImplemented";
})(StatusCode = exports.StatusCode || (exports.StatusCode = {}));
