// @xtool/logger/src/lib/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
class Logger {
    constructor(tag, opts) {
        var _a, _b;
        this.getMessageHeader = (arg) => {
            let message = `${this.tag}::${arg}`;
            if (this.withTimestamp) {
                message = `${this.tag}::${new Date().toISOString()}::${arg}`;
            }
            return message;
        };
        this.log = (arg, ...args) => {
            let message = this.getMessageHeader(arg);
            console.log(message, ...args);
        };
        this.debug = (arg, ...args) => {
            if (this.stage !== 'prod') {
                let message = this.getMessageHeader(arg);
                console.log(message, ...args);
            }
        };
        this.error = (arg, ...args) => {
            let message = this.getMessageHeader(arg);
            console.log(message, ...args);
        };
        this.tag = tag;
        this.stage = (_a = process.env.STAGE) !== null && _a !== void 0 ? _a : 'dev';
        this.withTimestamp = (_b = opts === null || opts === void 0 ? void 0 : opts.withTimestamp) !== null && _b !== void 0 ? _b : false;
    }
}
exports.Logger = Logger;
